File: matlab/src/matlab/+arrow/+array/Array.m
% Licensed to the Apache Software Foundation (ASF) under one or more
% contributor license agreements.  See the NOTICE file distributed with
% this work for additional information regarding copyright ownership.
% The ASF licenses this file to you under the Apache License, Version
% 2.0 (the "License"); you may not use this file except in compliance
% with the License.  You may obtain a copy of the License at
%
%   http://www.apache.org/licenses/LICENSE-2.0
%
% Unless required by applicable law or agreed to in writing, software
% distributed under the License is distributed on an "AS IS" BASIS,
% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
% implied.  See the License for the specific language governing
% permissions and limitations under the License.
    
classdef (Abstract) Array < matlab.mixin.CustomDisplay & ...
                            matlab.mixin.Scalar
% arrow.array.Array

    properties (GetAccess=public, SetAccess=private, Hidden)
        Proxy
    end

    properties(Dependent, SetAccess=private, GetAccess=public)
        NumElements
        Valid % Validity bitmap
        Type(1, 1) arrow.type.Type
    end
    
    methods
        function obj = Array(proxy)
            arguments
                proxy(1, 1) libmexclass.proxy.Proxy
            end
            obj.Proxy = proxy;
        end

        function numElements = get.NumElements(obj)
            numElements = obj.Proxy.getNumElements();
        end

        function validElements = get.Valid(obj)
            validElements = obj.Proxy.getValid();
        end

        function matlabArray = toMATLAB(obj)
            matlabArray = obj.Proxy.toMATLAB();
        end

        function type = get.Type(obj)
            typeStruct = obj.Proxy.getType();
            traits = arrow.type.traits.traits(arrow.type.ID(typeStruct.TypeID));
            proxy = libmexclass.proxy.Proxy(Name=traits.TypeProxyClassName, ID=typeStruct.ProxyID);
            type = traits.TypeConstructor(proxy);
        end

        function validate(obj, opts)
             arguments
                obj
                opts.Mode(1, 1) arrow.array.ValidationMode = arrow.array.ValidationMode.Minimal
             end

             if opts.Mode == arrow.array.ValidationMode.None
                 id = "arrow:array:InvalidValidationMode";
                 msg = "Invalid Mode. Mode must be ""Minimal"" or ""Full"".";
                 error(id, msg);
             end
             obj.Proxy.validate(struct(ValidationMode=uint8(opts.Mode)));
        end
    end

    methods (Access = private)
        function str = toString(obj)
            str = obj.Proxy.toString();
        end
    end

    methods (Access=protected)
        function header = getHeader(obj)
            name = matlab.mixin.CustomDisplay.getClassNameForHeader(obj);
            numElements = obj.NumElements;
            % TODO: Add NumValid and NumNull as properties to Array to
            % avoid materializing the Valid property. This will improve
            % performance for large arrays.
            numNulls = nnz(~obj.Valid);
            header = arrow.array.internal.display.getHeader(name, numElements, numNulls);
        end

        function displayScalarObject(obj)
            disp(getHeader(obj));
            if obj.NumElements > 0
                disp(toString(obj) + newline);
            end
        end
    end

    methods
        function tf = isequal(obj, varargin)
            narginchk(2, inf);
            tf = false;
            % Extract each array's proxy ID
            proxyIDs = zeros(numel(varargin), 1, "uint64");
            for ii = 1:numel(varargin)
                array = varargin{ii};
                if ~isa(array, "arrow.array.Array")
                    % Return early if array is not a arrow.array.Array
                    return;
                end
                proxyIDs(ii) = array.Proxy.ID;
            end
            % Invoke isEqual proxy object method
            tf = obj.Proxy.isEqual(proxyIDs);
        end

        function export(obj, cArrowArrayAddress, cArrowSchemaAddress)
            arguments
                obj(1, 1) arrow.array.Array
                cArrowArrayAddress(1, 1) uint64
                cArrowSchemaAddress(1, 1) uint64
            end
            args = struct(...
                ArrowArrayAddress=cArrowArrayAddress,...
                ArrowSchemaAddress=cArrowSchemaAddress...
            );
            obj.Proxy.exportToC(args);
        end
    end

    methods (Hidden)
        function array = slice(obj, offset, length)
            sliceStruct = struct(Offset=offset, Length=length);
            arrayStruct = obj.Proxy.slice(sliceStruct);
            traits = arrow.type.traits.traits(arrow.type.ID(arrayStruct.TypeID));
            proxy = libmexclass.proxy.Proxy(Name=traits.ArrayProxyClassName, ID=arrayStruct.ProxyID);
            array = traits.ArrayConstructor(proxy);
        end
    end

    methods (Static)
        function array = import(cArray, cSchema)
            arguments
                cArray(1, 1) arrow.c.Array
                cSchema(1, 1) arrow.c.Schema
            end
            importer = arrow.c.internal.ArrayImporter();
            array = importer.import(cArray, cSchema);
        end
    end
end

File: ruby/red-arrow/lib/arrow/jruby/array.rb
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

require_relative "data-type"

module Arrow
  class Array
    VectorAppender = org.apache.arrow.vector.util.VectorAppender
    VectorEqualsVisitor = org.apache.arrow.vector.compare.VectorEqualsVisitor

    attr_reader :vector

    def initialize(vector)
      @vector = vector
    end

    def ==(other_array)
      return false unless other_array.is_a?(self.class)
      VectorEqualsVisitor.vector_equals(@vector, other_array.vector)
    end

    def null?(i)
      @vector.null?(i)
    end

    def get_value(i)
      @vector.get_object(i)
    end

    def to_s
      @vector.to_s
    end

    def inspect
      super.sub(/>\z/) do
        " #{to_s}>"
      end
    end

    def close
      @vector.close
    end

    def length
      @vector.value_count
    end

    def value_data_type
      self.class::ValueDataType.new
    end

    def values
      each.to_a
    end

    def cast(other_value_data_type)
      other_value_data_type.build_array(to_a)
    end

    def is_in(values)
      raise NotImplementedError
    end

    def concatenate(other_arrays)
      total_size = length + other_arrays.sum(&:length)
      vector = self.class::Vector.new("", Arrow.allocator)
      vector.allocate_new(total_size)
      appender = VectorAppender.new(vector)
      @vector.accept(appender, nil)
      other_arrays.each do |other_array|
        other_array.vector.accept(appender, nil)
      end
      self.class.new(vector)
    end
  end

  class Int8Array < Array
    Vector = org.apache.arrow.vector.SmallIntVector
    ValueDataType = Int8DataType
  end

  class Int32Array < Array
    Vector = org.apache.arrow.vector.IntVector
    ValueDataType = Int32DataType
  end

  class FixedSizeBinaryArray < Array
  end

  class StructArray < Array
    def fields
      raise NotImplementedError
    end
  end
end
